src/oop/programs_with_multiple_classes/hard/q3/MainTest.java

```diff
@@ -0,0 +1,364 @@
+
+        try {
+            Book b = new Book(5, "fiction");
+        } catch (Error e) {
+            fail("Your program does not contain the " + bookClass + " class.");
+        }
+
+        assertTrue(bookcase.hasField(shelf), String.format(missingAttributeMessage, shelf));
+        assertTrue(bookcase.hasField(shelf, Book[].class), String.format(incorrectAttributeTypeMessage, shelf));
+        assertTrue(bookcase.hasModifier(shelf, "private"), String.format(incorrectAttributeModifierMessage, shelf));
+    }
+
+    @Test
+    public void bookcaseClassHasCorrectConstructors() {
+        try {
+            Object[][] arguments = new Object[][]{
+                    {5, int.class},
+                    {"fiction", String.class}
+            };
+            Object bookInstance = book.createInstance(arguments);
+        } catch (Error e) {
+            fail("Your program does not contain the " + bookClass + " class.");
+        }
+        String missingConstructorMessage = "Your " + bookcaseClass + " class does not have a required constructor.";
+        String incorrectConstructorModifierMessage =
+                "One of the constructors in your " + bookcaseClass + " class does not have the correct modifiers.";
+        Class<?>[] constructorClasses = new Class[]{
+                int.class
+        };
+        assertTrue(bookcase.hasConstructor(constructorClasses), missingConstructorMessage);
+        assertTrue(bookcase.hasModifier(constructorClasses, "public"), incorrectConstructorModifierMessage);
+    }
+
+    private static Stream<Arguments> bookcaseInputProvider() {
+        return Stream.of(
+                Arguments.of(4, 4),
+                Arguments.of(5, 5),
+                Arguments.of(11111, 11111),
+                Arguments.of(2, 2),
+                Arguments.of(0, 0),
+                Arguments.of(-3, 0),
+                Arguments.of(-1, 0),
+                Arguments.of(-8274, 0)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookcaseInputProvider")
+    public void bookcaseConstructorInitializesValuesCorrectly(int sizeCapacity, int result) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {sizeCapacity, int.class}
+        };
+        Object bookcaseInstance = bookcase.createInstance(arguments);
+        _assertEquals(result, bookcase.getFieldValue(bookcaseInstance, this.sizeCapacity),
+                "Your " + bookcaseClass + " constructor does not correctly initialize the " + this.sizeCapacity + " attribute.");
+        try {
+            Object[][] arguments = new Object[][]{
+                    {5, int.class},
+                    {"fiction", String.class}
+            };
+            Object bookInstance = book.createInstance(arguments);
+        } catch (Error e) {
+            fail("Your program does not contain the " + bookClass + " class.");
+        }
+        Book[] b = new Book[result];
```

Comment:
Don't do this. ObjectTest has a createArray method which will create an array of the book class without you having to call it. This will just throw a compilation error if there is no book class.